Migrating all binaries to the binaries repo

vedpatwardhan opened the pull request:
This PR removes the binaries from the `ivy` repo which would instead be stored on a dedicated repo named `binaries` to enable pulling only relevant binaries to a particular user's interpreter, OS and architecture.

vedpatwardhan commented:
I've done the following,

1. Removed all binaries from the `ivy` repo and added them instead to the [binaries](https://github.com/unifyai/binaries) repo with the naming convention same as for pypi wheels for consistency of naming. Also added a `binaries.json` to `ivy` for the skeleton structure. We'd need to add more binaries for different architectures and also different python versions as the current ones are only for python 3.9 and 3.10, but @vaatsalya123 could you please take a look at this part of the changes?
2. Updated the `setup.py` to download the binaries corresponding to a particular architecture with the same preference list of names of binaries as the names of wheels used by pypi (with `sys.tags()` being that preference list) with a spinner, a progress bar and so on. I'm not sure if it's as clean as it can be, @guillesanbri could you please take a look at the changes to `setup.py`?
3. In the rare case of someone wanting to clone repo, checkout to an older release (through the tag) and pulling the binaries corresponding to that release, I've added a "VERSION" environment variable they can set. But for this to be useful, we'd first need to ensure that we also add tags to the `binaries` repo when we make our releases. @RashulChutani could you please take a look at the changes made to `pypi.yml`?

Happy to know about improvements or clarify about any of these changes, thanks!

guillesanbri commented on setup.py:
```diff
@@ -18,13 +18,71 @@
 import setuptools
 from setuptools import setup
 from pathlib import Path
+from packaging import tags
```
had to install this manually

vedpatwardhan replied:
replaced it with a built-in alternative, thanks for pointing that out!

guillesanbri commented on setup.py:
```diff
@@ -18,13 +18,71 @@
+from packaging import tags
+import os
+import json
+import requests
```
same with requests

guillesanbri commented on setup.py:
```diff
@@ -18,13 +18,71 @@
+import requests
+import itertools
 import re
+from tqdm import tqdm
```
and same with tqdm

guillesanbri commented on setup.py:
```diff
@@ -18,13 +18,71 @@
+                print()
+                print("Downloading binaries ...")
+                pbar = tqdm(total=len(paths))
+            with open(path, "wb") as f:
```
can't pass over this point, I get:
```
          with open(path, "wb") as f:
      FileNotFoundError: [Errno 2] No such file or directory: 'ivy/compiler/utils/III.so'
```

Seems to work if I create the utils dir manually, so I'm assuming we are missing `os.makedirs(os.path.dirname(path), exist_ok=True)` before `open`

vedpatwardhan replied:
done, thanks for pointing that out!

guillesanbri commented on setup.py:
```diff
@@ -18,13 +18,71 @@
+                f.write(r.content)
+            terminate = path == paths[-1]
+            pbar.update(1)
+        else:
```
maybe worth adding an error message if the connection hasn't been successful :)

guillesanbri replied:
Ah got it yeah, this works fine then, thanks Ved! :+1:

guillesanbri reviewed (commented):
Left a few comments, I tried installing in a fresh environment, which I assume is the cause of the missing packages, will need to sort this out nonetheless :sweat_smile: 

I think the `makedirs` call should fix the other problem I was facing. I tried on a fresh conda environment, let me know if something else should be specially tested :+1:

vedpatwardhan commented:
done with the changes, please take a look whenever you're free, thanks!

guillesanbri commented on setup.py:
```diff
@@ -18,13 +18,72 @@
+        except request.HTTPError:
+            break
+
+for path in paths:
+    if not os.path.exists(path):
+        print(
+            f"Could not download {path}. Check out the unifyai/binaries README for"
+            " the supported configurations!"
+        )
```
lgtm!

vedpatwardhan commented on requirements/requirements.txt:
```diff
@@ -7,6 +7,8 @@ packaging
 requests
 pyvis
 dill
+tqdm
```
reverted the changes to `requirements` and `optional` files 👍
actually `tqdm` is being used in the `determine_test_coverage.py` file, probably @RashulChutani could comment on it's usability 😅

guillesanbri replied:
ah missed that, true!

vedpatwardhan replied:
no worries!

vedpatwardhan commented:
> 1. Removed all binaries from the `ivy` repo and added them instead to the [binaries](https://github.com/unifyai/binaries) repo with the naming convention same as for pypi wheels for consistency of naming. Also added a `binaries.json` to `ivy` for the skeleton structure. We'd need to add more binaries for different architectures and also different python versions as the current ones are only for python 3.9 and 3.10, but @vaatsalya123 could you please take a look at this part of the changes?
> 3. In the rare case of someone wanting to clone repo, checkout to an older release (through the tag) and pulling the binaries corresponding to that release, I've added a "VERSION" environment variable they can set. But for this to be useful, we'd first need to ensure that we also add tags to the `binaries` repo when we make our releases. @RashulChutani could you please take a look at the changes made to `pypi.yml`?

Hey @vaatsalya123 and @RashulChutani, could you please take a look? Given it's directly linked to the binaries, I'm refraining from merging without multiple opinions. Thanks!

vedpatwardhan commented:
> Overall, it looks good to me. However, the binaries code is better suited to be in the release workflow instead of pypi.yml, as that is dedicated for PyPI deployment.

Thanks for the review, I've basically added a `binaries.yml` for adding the tags to the binaries repo and added it as a job to `release.yml`, happy to know your thoughts 😄

vaatsalya123 commented:
@vedpatwardhan All look's perfect to me , Just so I'm sure did you test the full pipeline in action to confirm it's working or are we yet to do this?

vedpatwardhan commented:
I'll get this merged for now, let me know if there's anything to be updated with the `binaries.yml` workflow @RashulChutani I'll get that fixed. Ofc given that the `binaries.yml` workflow wouldn't be used until we make a release, it would lead to any immediate problems so merging the PR for now. Thanks all 😄
